### ISSUE TITLE
Webpack bundling with node target fails

### ISSUE THREAD

**Issue by samelliottdlt** [openpgpjs/openpgpjs#896]
Here is a codesandbox: https://codesandbox.io/s/0031z8qymn

To reproduce run the following commands in the terminal:

`npm run build`
`node dist/index.js`

This will error with `Error: Cannot find module 'stream'`

It seems that this is failing due to [the nodeRequire function](https://github.com/openpgpjs/openpgpjs/blob/master/src/util.js#L596) not being statically analyzable by webpack. Within that same file there are a few functions to require node built-ins that call the nodeRequire function instead of using require directly which would be statically analyzable.

I feel like ideally the entry point of the package should be within src/index.js rather than dist/openpgp.js. Module bundlers would then be able to handle bundling the code and users can add any prerequisite polyfills.

**Comment by twiss** [comment]
Hey :wave: I agree that this would be a good idea. Feel free to take a stab at it, if you like.

However, `nodeRequire` doesn't do anything if openpgp.js detects that it's not running in Node, so its existence, and the fact that it's not statically analyzable, shouldn't be an issue. In Gruntfile.js, there's a list of modules that *shouldn't* be included in the bundle, and stream is one of them. So in fact, you might *fix* this issue by replacing `require('stream')` with `util.nodeRequire('stream')`.

**Comment by samelliottdlt** [comment]
@twiss I was able to fix this issue on our end by configuring webpack to understand this particular dependency. Ideally, this configuration wouldn't be necessary.

I agree that it being not statically analyzable shouldn't be an issue but due to the way webpack and other module bundlers are implemented it (unfortunately) is. Webpack likes to overload require to its own implementation of require and trying to require a module that it didn't find when statically analyzing the code results in an error.

I'll try to take a stab at it as time permits. It'll probably require using a module bundler like rollup and generate bundles for different build targets: es5, es2015 and umd. This way the package can support all the current supported environments. Lets see how it goes :)

**Comment by adhooaa** [comment]
@samelliottdlt 
I just ran into this myself. Can you please share your webpack config workaround to resolve the dependency?

**Comment by samelliottdlt** [comment]
@adhooaa Here are the relevant bits you probably need. Our webpack config is kinda large-ish and contains a lot of stuff so I cut out all the noise.

```
resolve: {
        alias: {
          openpgp: 'openpgp/src/index.js'
        }
},
plugins: [
          new webpack.DefinePlugin({
            'global.GENTLY': false
          }),
          new webpack.ContextReplacementPlugin(
              /openpgp/,
              '',
              {
                crypto: 'crypto',
                stream: 'stream',
                buffer: 'buffer',
                zlib: 'zlib',
                util: 'util'
              }
          )
]

```

What we are doing here is overwriting the entry point to instead go to src/index.js instead of dist/openpgp.js. We then tell webpack using the ContextReplacePlugin how to resolve the node built-ins. You will also need to npm i a few dependencies. I recommend installing those as they come up.
[Note that some dependencies are installed using github urls rather than using the npm registry](https://github.com/openpgpjs/openpgpjs/blob/master/package.json#L75)  and you will need to install these using those github urls! Let me know how it goes :)

I am working on a fix on a branch here: https://github.com/samelliottdlt/openpgpjs

Feel free to take a look and maybe contribute!

**Comment by studds** [comment]
What's the status of this branch @samelliottdlt? Anything in particular you need help with? @twiss  any comments on this approach? Would you be open to merging this?

**Comment by twiss** [comment]
Hey @studds and @samelliottdlt :wave:
```
          new webpack.ContextReplacementPlugin(
              /openpgp/,
              '',
              {
                crypto: 'crypto',
                stream: 'stream',
                buffer: 'buffer',
                zlib: 'zlib',
                util: 'util'
              }
          )
```
If I understand it correctly, this tells webpack to include these modules in the bundle, right? Ideally, we want the exact opposite; we want to exclude them. OpenPGP.js should still work without them, and only `require()`s them when it detects it's running on Node.js.
If there's something we can merge that instructs webpack to do so when building OpenPGP.js as a dependency, I'd be happy to do so.

**Comment by samelliottdlt** [comment]
Hey @studds I haven't worked on that branch in quite a while. Other responsibilities have kept me from digging deeper on this. Feel free to fork it and take a stab at it.

@twiss Correct kinda. It tells webpack to include these in its dependency graph so it can resolve to the node built ins during runtime. I'm going off memory but here is what I remember: in the openpgpjs source code is doesn't use a regular call to require require and instead uses a function called nodeRequire which is not statically analyzable by webpack. This results in webpack not including those node built-ins in it's dependency graph. Webpack during runtime overloads the calls to requires and sees a require to something that it find during static analysis and throws an error.

I think this can be fixed by not using the nodeRequire function and instead using a regular require call. There may be other issues to fix that I don't remember off hand but that may be a good first step in the right direction.

**Comment by samelliottdlt** [comment]
It would look something like this:

replace `nodeRequire('zlib')` with `if (checkIfRunningInNode) require('zlib')`

**Comment by twiss** [comment]
> It tells webpack to include these in its dependency graph so it can resolve to the node built ins during runtime.

Thanks for the explanation! That sounds fine then.

> Webpack during runtime overloads the calls to requires and sees a require to something that it did not find during static analysis and throws an error.

To be clear: are you trying to create a bundle for use on Node.js or for in the browser? If you're running on Node, I understand the error, but I'm not sure why you need webpack in that case? Have you tried whether the bundle works (with or without this configuration) in the browser?

> I think this can be fixed by not using the nodeRequire function and instead using a regular require call.

At least when using browserify, this would cause polyfills of the modules to be included in the bundle, which we don't want:

https://github.com/openpgpjs/openpgpjs/blob/8585ad892498806d8ef4654619b9a9074bfe028c/src/util.js#L597-L606

I assumed webpack would do something similar? (https://webpack.js.org/configuration/node/#other-node-core-libraries)

**Comment by samelliottdlt** [comment]
I've only run our bundle in node. My workplace deploys node bundles to micro services and bundle size matters here. I don't think bundling server code is all that uncommon but should be a sizeable niche. I can try giving it a shot in running in the browser - I expect it to work well but you never know.

As an aside, seems like turning on the --no-builtins flag on browserify will make browserify not include the polyfills. Per the browserify docs:

```
  --no-builtins

    Turn off builtins. This is handy when you want to run a bundle in node which
    provides the core builtins.
```

**Comment by pwhitfield-cashwerkz** [comment]
is there any resolution to this issue? Encountered with project using webpack to deploy serverless project to AWS lambda on Node 12.x

**Comment by tripodsan** [comment]
the last time I tested, it still didn't work.
feel free to use my fork: https://github.com/tripodsan/openpgpjs/tree/v4.10.1-node8-3
npm: https://www.npmjs.com/package/@tripod/openpgp

**Comment by twiss** [comment]
Hey :wave: This PR should make the build work with webpack: https://github.com/openpgpjs/openpgpjs/pull/1095
However, since it's on the v5 branch, there are API changes as well, documented here: https://github.com/openpgpjs/openpgpjs/wiki/V5-Changelog
There will probably be further API changes before we release v5, but if you wish, you could try it already.

**Comment by tripodsan** [comment]
@twiss Thanks for the work. I'll look into it. 

One other issue I had with the original repository were all the git-referencing dependencies, like: https://github.com/openpgpjs/openpgpjs/pull/1095/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R48

those are really a pain when _npm-installing_ the library, since resolving/downloading those are considerably slower than fetching the packages from npm. it would be really nice, if you could create proper NPM packages for your custom packages. this also makes openpgpjs look more mature...right now, it looks like cobbled together.

**Comment by tripodsan** [comment]
Hi @twiss, what's the status regarding a v5 release?

**Comment by twiss** [comment]
Hey @tripodsan, thanks for trying it out! I will look into publishing the dependencies on npm.

There are still some API changes I want to make. Also, we might want to wait on RFC4880bis or GnuPG 2.3 to be released, given some of the new features that are used in this version. So it might still take quite a while, honestly.

**Comment by tripodsan** [comment]
> Perhaps I might publish a prerelease to npm, though. Would that help?

that would be great. especially, if the dependencies are also released. thanks

**Comment by TikiTDO** [comment]
I found another workaround for this while waiting for the next version. This works if you are able to upload multiple files, which makes it ideal for any tools that deploy lambda functions from directories.

```js
// webpack config
module.exports = {
  // ...,
  externals: {
    openpgp: "commonjs2 ./openpgp.min"
  },
  plugins: [
    new CopyWebpackPlugin({
      patterns: [
        {
          from: path.join(
            ROOT_PATH,
            "node_modules/openpgp/dist/openpgp.min.js",
          ),
        },
      ],
    }),
}
```

As long as both the build package is uploaded intact it should run without problems correctly.

**Comment by AG-Labs** [comment]
Hi @TikiTDO I have tried the above but am running into errors when I try, is there anything else that is required?

My current error in the lambda is         

`"Runtime.ImportModuleError: Error: Cannot find module './openpgp.min'",`

**Comment by TikiTDO** [comment]
@AG-Labs Looks like you did not actually upload `openpgp.min.js` to your lambda distribution. The above webpack configuration 
will copy the lib as a separate file to the output directory, in addition to your actual lambda code. After that it's up to you to upload both your code, as well as the lib file to AWS.

**Comment by AG-Labs** [comment]
ah yes seen it in the wrong place in my repo. Might not work as im using cdk so i think that only accepts single files. Might have to try the other solutions higher up this chain.
thank you @TikiTDO

**Comment by TikiTDO** [comment]
@AG-Labs If you're using the AssetCode CDK uploader then it should be fairly simple. Here's a snipped based on my config:

```ts
const lambdaSecurePath = path.resolve(`${ROOT_PATH}/backend/secure`)

const buildPath = path.resolve(`${ROOT_PATH}/build`)
const lambdaSecureBuildPath = path.resolve(`${buildPath}/secure/`)

const secureWebpackConfig: webpack.Configuration = {
  entry: {
    // Used by openpgp.js
    asn1: `${ROOT_PATH}/node_modules/asn1.js/lib/asn1.js`,
    example: `${lambdaSecurePath}/example.ts`,
  },
  // Rewrites the import to use the minified version
  externals: [
    (context, requests, callback) => {
       const pgpRequest = path
         .join(context, requests)
         .match(/\bopenpgp\b/)?.[0]

       if (pgpRequest) {
         callback(null, "commonjs2 ./openpgp.min")
       } else {
         callback()
       }
     },
  ],
  module: {
    rules: [
      {
        // Ensures webpack doesn't try to pack openpgp.js
        exclude: (input) => {
          return !!input.match(/\b(openpgp)(\/.*)?/)
        },
        test: /\.(js|jsx|ts|tsx)$/,
        use: {
          // Your loader may vary
          loader: "babel-loader",
          options: {
            cacheDirectory: true,
            presets: [
              "@babel/preset-typescript",
              ["@babel/preset-env", { targets: { node: "10" } }],
            ],
          },
        },
      },
    ],
    plugins: [
      new CopyWebpackPlugin({
        patterns: [
          {
            from: path.join(
              ROOT_PATH,
              "node_modules/openpgp/dist/openpgp.min.js",
            ),
          },
        ],
      }),
    ],
  },
  output: {
    filename: "[name].js",
    libraryTarget: "umd" as const,
    path: lambdaSecureBuildPath,
  },
}

webpack(secureWebpackConfig).run((_error, stats) => {
  // Handle errors and such...

  new lambda.Function(this, "example", {
    code: new lambda.AssetCode(lambdaSecureBuildPath),
    handler: "example.handler",
    runtime: lambda.Runtime.NODEJS_10_X,
    timeout: cdk.Duration.seconds(90),
  })
}
```

**Comment by brander-john** [comment]
@twiss This doesn't exist anymore, is there another commit or branch I can point to?

> ```
>     "openpgp": "github:twiss/openpgpjs#90d83ffaea110e4eefb76f91e237bbfc4d43d15b",
> ```

```
[1/4] 🔍  Resolving packages...
error Couldn't find match for "90d83ffaea110e4eefb76f91e237bbfc4d43d15b"
```

**Comment by TheMechanic** [comment]
Hello, just had the same issue. You just have to replace `import openpgp from 'openpgp'` by `import openpgp from 'openpgp/dist/openpgp.min.js'` (to embed as few weight as possible - but almost 600Ko, ouch -), then, add to your webpack config 'openpgp/dist/openpgp.min.js' to externals !

**Comment by twiss** [comment]
This should be fixed by #1047 (in the upcoming v5 release).
